### DISCUSSION TITLE
Have separate function tracks for each function in Animation Player

### DISCUSSION THREAD

**Discussion by SanNull** [godotengine/godot-proposals#9770] [Editor]
Calling a Node's functions in the Animation Timeline is really handy and it also eases the development process. However, it may  become hard to read and maintain if you call many of them, since they're all in the same track.

As an example, let's say you're animating a cutscene and would like to call the function `dialogueClip` that takes a String as a parameter. As it stands now, calling it multiple times looks like this:

![animation_player1](https://github.com/godotengine/godot-proposals/assets/124409014/0f1d6139-82e0-4889-a025-b2db5458bfb9)

By contrast, the properties found in _Icon_ don't look messy at all. This is a simple example that could be circumvented, but the more functions you add, the longer their names, the longer the arguments passed into them, the more pointless it is to have the functions' name and arguments displayed directly on the Timeline. So instead of that, the suggested change is to separate them into their own tracks, similar to how it is done with properties:

![beatboxingpuppy](https://github.com/godotengine/godot-proposals/assets/124409014/3c2d074d-d7f0-4fd1-8693-fd0ff910f7b7)

Arguments will still be passed through the Inspector, with the advantage of not polluting the Timeline:

![as](https://github.com/godotengine/godot-proposals/assets/124409014/343f27b2-675c-46b1-89f3-1624425673f2)

Having tracks for each function improves readability and makes the Animation player more organized to animate with.

**Comment by TokageItLab** [comment]
If you actually separate the functions into individual tracks, the number of track caches will increase by the number of functions, so at least the separation must be done in only for visually in the GUI with implementations of folding, etc., without increasing track count.

Also, this is a known problem, but there is no easy way to control the order of multiple functions/values applying at the same key time, so while it may be possible to prohibit placing functions at the same time on the GUI, if you want to process them properly, we will need to make fundamental rework to control the order of processing keys at the same time.

**Reply by SanNull** [reply]
I see, maybe what could be done to declutter the track is to shorten part of function name with '...' whenever it's overlapping with another key, although I'm not sure how to go about it. This sounds more concise and doesn't create extra complexity to both the user and contributor. 

Edit: After testing a bit, it seems that it wouldn't solve the issue in a satisfactory way, since you might as well just keep the function names and args hidden, perhaps the name shows up only when hover/click on the key, but I wonder if this would be a viable solution.

**Reply by TokageItLab** [reply]
I think it would be reasonably straightforward and feasible to implement the ability to hide argument or omit text in the editor.

**Reply by SanNull** [reply]
Was able to add the [feature](https://github.com/godotengine/godot/pull/96421), I will be closing this discussion now.
